File: Number Theory/sieve.cpp
constexpr int N = 1000000;

bool prime[N + 1];

/* O(N * Log(Log(N))). */
vector<int> sieve(int n) {
    vector<int> p;

    // Initializing.
    memset(prime, true, sizeof(prime));
    prime[0] = prime[1] = false;

    // For every number i from 2 to N.
    for (int i = 2; i <= n; i++) {
        // If current number is prime.
        if (prime[i]) {
            p.push_back(i);

            // Marking its multiples as not prime.
            for (int j = i * i; j <= n; j += i) {
                prime[j] = false;
            }
        }
    }

    return p;
}

File: Number Theory/sieve.py
# O(N) - Returns the prime numbers up to n.
def sieve(n):
	# Initializing.
	prime = [True for i in range(n + 1)]
	prime[0] = prime[1] = False
	p = []

	# For every number i.
	for i in range(2, n + 1):
		# If current number is prime.
		if prime[i]:
			p.append(i)

			# Marking its multiples as not prime.
			j = i * i

			while j <= n:
				prime[j] = False
				j += i

	return p